File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/MegaSpeechBubble.ts
import { FairyGUI, SuperText } from "csharp";
import {View_SpeechBubble} from "mods/ModTheSpire/Scripts/Gen/View/ModTheSpire_Effect";
import UIHelper from "mods/ModTheSpire/Scripts/UI/UIHelper";
import Color from "mods/ModTheSpire/Scripts/DataDefine/Color";
import {AbstractEffect} from "mods/ModTheSpire/Scripts/Effect/AbstractEffect";

export class MegaSpeechBubble extends AbstractEffect{
    private Instance : FairyGUI.GComponent;
    private Bg : FairyGUI.GImage;
    private SuperText:SuperText;
    public constructor(private X:number,private Y:number,private Msg:string,private Interval?:number,private FlipX? : boolean) {
        super();
        this.Instance = <FairyGUI.GComponent>View_SpeechBubble.CreateInstance();
        this.Bg = this.Instance.GetChild("Bg").asImage;
        if(FlipX)
            this.Bg.flip = FairyGUI.FlipType.Horizontal;
        let graph = this.Instance.GetChild("Holder").asGraph;
        this.SuperText = UIHelper.CreateSuperText();
        let wrap = new FairyGUI.GoWrapper(this.SuperText.gameObject);
        wrap.SetScale(100,100)
        this.SuperText.color = Color.Gray.UnityColor();
        this.SuperText.text = `{speechsize}{offset}${Msg}{/offset}{/speechsize}`;
        this.SuperText.lineSpacing = -50;
        this.SuperText.Width = 2;
        graph.SetNativeObject(wrap);
        this.Instance.GetTransition("t0").Play();
    }
    Update() {
        super.Update();
        this.SuperText.alpha = this.Bg.alpha;
        if(this.Duration >= this.Interval)
            this.IsDone = true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/MegaSpeechBubble.cs
using FairyGUI;
using Febucci.UI;
using ModTheSpire.DataDefine;
using ModTheSpire.ModTheSpire_Effect;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class MegaSpeechBubble : AbstractEffect
    {
        private float mX;
        private float mY;
        private string mMsg;
        private int mInterval;
        private bool mFlipX;
        private GComponent mInstance;
        private TextAnimator mSuperText;
        private GImage mBg;
        public MegaSpeechBubble(float x, float y, string msg, int interval, bool flipX)
        {
            mX = x;
            mY = y;
            mMsg = msg;
            mInterval = interval;
            mFlipX = flipX;
            mInstance = View_SpeechBubble.CreateInstance();
            mBg = mInstance.GetChild("Bg").asImage;
            if(mFlipX)
                mBg.flip = FlipType.Horizontal;
            var graph = mInstance.GetChild("Holder").asGraph;
            mSuperText = Resources.Load<GameObject>("TMP").GetComponent<TextAnimator>();
            var wrap = new GoWrapper(mSuperText.gameObject);
            wrap.SetScale(100, 100);
            var textMeshPro = mSuperText.tmproText;
            textMeshPro.color = MTSColor.Gray;
            textMeshPro.text = "{speechsize}{offset}" + mMsg + "{/offset}{/speechsize}";
            textMeshPro.lineSpacing = -50;
            var size = textMeshPro.rectTransform.sizeDelta;
            size.x = 2;
            textMeshPro.rectTransform.sizeDelta = size;
            graph.SetNativeObject(wrap);
            mInstance.GetTransition("t0").Play();
        }

        public override void Update()
        {
            base.Update();
            mSuperText.tmproText.alpha = mBg.alpha;
            if(Duration >= mInterval)
                Finish();
        }

        public override void Dispose()
        {
            base.Dispose();
            mInstance.Dispose();
        }
    }
}